src/query/functions/binary/or.go

```diff
@@ -74,34 +100,48 @@ func makeOrBlock(
 		return nil, err
 	}
 
-	if err = appendValuesAtIndices(missingIndices, rIter, builder); err != nil {
+	if rIter.Next() {
+		return nil, errLExhausted
+	}
+
+	if err = rIter.Err(); err != nil {
 		return nil, err
 	}
 
 	return builder.Build(), nil
 }
 
-// missing returns the slice of rhs indices for which there are no corresponding
-// indices on the lhs. It also returns the metadatas for the series at these
-// indices to avoid having to calculate this separately
-func missing(
+// mergeIndices returns a slice that maps rhs series to lhs series.
+//
+// NB(arnikola): if the series in the rhs does not exist in the lhs, it is
+// added after all lhs series have been added.
+// This function also combines the series metadatas for the entire block.
+func mergeIndices(
 	matching *VectorMatching,
 	lhs, rhs []block.SeriesMeta,
 ) ([]int, []block.SeriesMeta) {
 	idFunction := HashFunc(matching.On, matching.MatchingLabels...)
```

Comment:
Leaving a self note more than anything, we do need to go revisit using `fnv` hash to guarantee uniqueness for map (there's actually a lot of collisions using `fnv`). The auto-gen map we have will use `xxhash` (faster than fnv) and ensure collisions are treated properly, also without allocating/copying the keys or values if desired.

Reply:
Was sure we were looking at this earlier and were fine with it since that was what Prom was using, but looking into Prom's approach now, they use [xxHash](https://sourcegraph.com/github.com/prometheus/prometheus/-/blob/pkg/labels/labels.go?utm_source=VSCode-1.0.14#L152) so will update to match